src/result.rs

```diff
@@ -1,9 +1,74 @@
 use std::convert::From;
-use std::fmt::{Debug, Error};
+use std::fmt::{Debug, Display, Error};
 use std::{fmt, io};
 
 use thiserror::Error;
 
+/// Represents the position, as line and column, within a text format ion document.
+#[derive(Clone, Debug)]
+pub enum TextPosition {
+    /// No text position available; implies binary format is being read.
+    None,
+    /// The position represented as line and column.
+    LineColumn(usize, usize),
+}
+
+#[derive(Clone, Debug)]
+pub struct Position {
+    pub byte_offset: usize,
+    pub line_column: TextPosition,
+}
+
+impl Position {
+    fn with_offset(offset: usize) -> Self {
+        Position {
+            byte_offset: offset,
+            line_column: TextPosition::None,
+        }
+    }
+
+    fn with_line_info(offset: usize, line: usize, column: usize) -> Self {
+        Position {
+            byte_offset: offset,
+            line_column: TextPosition::LineColumn(line, column),
+        }
+    }
+}
+
+impl PartialEq for Position {
+    fn eq(&self, other: &Self) -> bool {
+        if other.byte_offset != self.byte_offset {
+            false
+        } else {
+            use TextPosition::*;
+            match (&other.line_column, &self.line_column) {
+                (None, None) => true,
+                (None, LineColumn(_, _)) | (LineColumn(_, _), None) => false,
+                (LineColumn(ol, oc), LineColumn(sl, sc)) => ol == sl && oc == sc,
+            }
+        }
+    }
+
+    fn ne(&self, other: &Self) -> bool {
+        !self.eq(other)
+    }
+}
+
+impl Eq for Position {}
```

Comment:
In Java, it's a best practice to implement `equals()` and `hashCode()`, and it seems like a principle that would be language-independent. Should there also be an `impl Hash for Position` here?